pkg/controllers/managedserviceaccount/log_serviceaccount_controller.go

```diff
@@ -0,0 +1,142 @@
+func newReconciler(mgr manager.Manager) *Reconciler {
+	return &Reconciler{
+		client:     mgr.GetClient(),
+		scheme:     mgr.GetScheme(),
+		logMcaName: helpers.LogManagedServiceAccountName,
+	}
+}
+
+// add adds a new Controller to mgr with r as the reconcile.Reconciler
+func add(mgr manager.Manager, r *Reconciler) error {
+	// Create a new controller
+	c, err := controller.New("log-managedServiceAccount-controller", mgr, controller.Options{Reconciler: r})
+	if err != nil {
+		return err
+	}
+
+	// watch ManagedCluster as the primary resource
+	err = c.Watch(source.Kind(mgr.GetCache(), &clusterv1.ManagedCluster{}), &handler.EnqueueRequestForObject{})
+	if err != nil {
+		return err
+	}
+
+	// watch log ManagedServiceAccount
+	err = c.Watch(source.Kind(mgr.GetCache(), &msav1beta1.ManagedServiceAccount{}), handler.EnqueueRequestsFromMapFunc(
+		handler.MapFunc(func(ctx context.Context, a client.Object) []reconcile.Request {
+			if a.GetName() != helpers.LogManagedServiceAccountName {
+				return []reconcile.Request{}
+			}
+
+			return []reconcile.Request{
+				{
+					NamespacedName: types.NamespacedName{
+						Name: a.GetNamespace(),
+					},
+				},
+			}
+		}),
+	))
+	if err != nil {
+		return err
+	}
+
+	return nil
+}
+
+func (r *Reconciler) Reconcile(ctx context.Context, req ctrl.Request) (ctrl.Result, error) {
+	cluster := &clusterv1.ManagedCluster{}
+	err := r.client.Get(ctx, req.NamespacedName, cluster)
+	if errors.IsNotFound(err) {
+		return ctrl.Result{}, nil
+	}
+	if err != nil {
+		return ctrl.Result{}, err
+	}
+	if !cluster.DeletionTimestamp.IsZero() {
+		return ctrl.Result{}, nil
+	}
+
+	addon := &addonv1alpha1.ManagedClusterAddOn{}
+	err = r.client.Get(ctx, types.NamespacedName{Name: helpers.MsaAddonName, Namespace: cluster.Name}, addon)
+	if errors.IsNotFound(err) {
+		return ctrl.Result{}, nil
+	}
+	if err != nil {
+		return ctrl.Result{}, err
+	}
+	if !addon.DeletionTimestamp.IsZero() {
+		return ctrl.Result{}, nil
+	}
+
+	requiredLogMsa := r.newLogMsa(cluster.Name)
+	logMsa := &msav1beta1.ManagedServiceAccount{}
+	err = r.client.Get(ctx, types.NamespacedName{Name: r.logMcaName, Namespace: cluster.Name}, logMsa)
+	if errors.IsNotFound(err) {
+		return ctrl.Result{}, r.client.Create(ctx, requiredLogMsa)
+	}
+	if err != nil {
+		return ctrl.Result{}, err
+	}
+
+	if apiequality.Semantic.DeepEqual(requiredLogMsa.Spec, logMsa.Spec) {
+		return ctrl.Result{}, nil
+	}
+	logMsa.Spec = requiredLogMsa.Spec
+	return ctrl.Result{}, r.client.Update(ctx, logMsa)
+}
+
+func (r *Reconciler) newLogMsa(clusterName string) *msav1beta1.ManagedServiceAccount {
+	return &msav1beta1.ManagedServiceAccount{
+		TypeMeta: metav1.TypeMeta{},
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      r.logMcaName,
+			Namespace: clusterName,
+		},
+		Spec: msav1beta1.ManagedServiceAccountSpec{
+			Rotation: msav1beta1.ManagedServiceAccountRotation{
+				Enabled:  true,
+				Validity: metav1.Duration{Duration: time.Minute * 365 * 24 * 60},
```

Comment:
TODO, 1 year is quite long for a token. Should consider a token like 24 hours or 1 week maximum

Reply:
updated to 1 week